File: content/tracks/dsa-typescript/concepts/07-queues-and-deques/queues-and-deques_test.ts
import { slidingWindowMaximum, RecentCounter, rottingOranges } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

// --- slidingWindowMaximum ---

test("slidingWindowMaximum: standard case k=3", () => {
  assertEqual(slidingWindowMaximum([1, 3, -1, -3, 5, 3, 6, 7], 3), [3, 3, 5, 5, 6, 7]);
});

test("slidingWindowMaximum: k=1 returns input", () => {
  assertEqual(slidingWindowMaximum([1, 2, 3], 1), [1, 2, 3]);
});

test("slidingWindowMaximum: k equals length", () => {
  assertEqual(slidingWindowMaximum([3, 1, 2], 3), [3]);
});

test("slidingWindowMaximum: decreasing sequence", () => {
  assertEqual(slidingWindowMaximum([5, 4, 3, 2, 1], 2), [5, 4, 3, 2]);
});

test("slidingWindowMaximum: single element", () => {
  assertEqual(slidingWindowMaximum([7], 1), [7]);
});

test("slidingWindowMaximum: all same values", () => {
  assertEqual(slidingWindowMaximum([2, 2, 2, 2], 2), [2, 2, 2]);
});

test("slidingWindowMaximum: negative values", () => {
  assertEqual(slidingWindowMaximum([-5, -3, -1, -2], 2), [-3, -1, -1]);
});

// --- RecentCounter ---

test("RecentCounter: first ping returns 1", () => {
  const rc = new RecentCounter();
  assertEqual(rc.ping(1), 1);
});

test("RecentCounter: pings within window", () => {
  const rc = new RecentCounter();
  rc.ping(1);
  rc.ping(100);
  rc.ping(3001);
  assertEqual(rc.ping(3002), 3);
});

test("RecentCounter: old pings expire", () => {
  const rc = new RecentCounter();
  rc.ping(1);
  rc.ping(2);
  assertEqual(rc.ping(3002), 2); // ping(1) is now outside [2, 3002]
});

test("RecentCounter: exact boundary included", () => {
  const rc = new RecentCounter();
  rc.ping(100);
  assertEqual(rc.ping(3100), 2); // 3100 - 3000 = 100, included
});

test("RecentCounter: just outside boundary excluded", () => {
  const rc = new RecentCounter();
  rc.ping(100);
  assertEqual(rc.ping(3101), 1); // 3101 - 3000 = 101 > 100, excluded
});

// --- rottingOranges ---

test("rottingOranges: standard case → 4", () => {
  assertEqual(rottingOranges([[2, 1, 1], [1, 1, 0], [0, 1, 1]]), 4);
});

test("rottingOranges: isolated fresh orange → -1", () => {
  assertEqual(rottingOranges([[2, 1, 1], [0, 1, 1], [1, 0, 1]]), -1);
});

test("rottingOranges: no fresh oranges → 0", () => {
  assertEqual(rottingOranges([[0, 2]]), 0);
});

test("rottingOranges: single fresh, single rotten adjacent → 1", () => {
  assertEqual(rottingOranges([[2, 1]]), 1);
});

test("rottingOranges: all fresh, no rotten → -1", () => {
  assertEqual(rottingOranges([[1, 1], [1, 1]]), -1);
});

test("rottingOranges: empty grid → 0", () => {
  assertEqual(rottingOranges([[0, 0], [0, 0]]), 0);
});

test("rottingOranges: already all rotten → 0", () => {
  assertEqual(rottingOranges([[2, 2], [2, 2]]), 0);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/07-queues-and-deques/queues-and-deques_test.py
import sys
from stub import sliding_window_maximum, RecentCounter, rotting_oranges

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# sliding_window_maximum
r = sliding_window_maximum([1, 3, -1, -3, 5, 3, 6, 7], 3)
test("swmax: standard k=3", r == [3, 3, 5, 5, 6, 7], f"got {r}")
r = sliding_window_maximum([1], 1)
test("swmax: single element", r == [1], f"got {r}")
r = sliding_window_maximum([1, 2, 3, 4, 5], 2)
test("swmax: ascending k=2", r == [2, 3, 4, 5], f"got {r}")
r = sliding_window_maximum([5, 4, 3, 2, 1], 3)
test("swmax: descending k=3", r == [5, 4, 3], f"got {r}")
r = sliding_window_maximum([9, 9, 9], 2)
test("swmax: all equal", r == [9, 9], f"got {r}")
r = sliding_window_maximum([2, 1, 5, 3, 6, 4, 8, 2], 4)
test("swmax: k=4 mixed", r == [5, 6, 6, 8, 8], f"got {r}")

# RecentCounter
rc = RecentCounter()
r = rc.ping(1)
test("counter: first ping", r == 1, f"got {r}")
r = rc.ping(100)
test("counter: second ping", r == 2, f"got {r}")
r = rc.ping(3001)
test("counter: third ping", r == 3, f"got {r}")
r = rc.ping(3002)
test("counter: fourth ping (first expires)", r == 3, f"got {r}")
rc2 = RecentCounter()
rc2.ping(0)
r = rc2.ping(3001)
test("counter: ping 0 still in window at 3001", r == 2, f"got {r}")
r = rc2.ping(3002)
test("counter: ping 0 expires at 3002", r == 2, f"got {r}")

# rotting_oranges
r = rotting_oranges([[2, 1, 1], [1, 1, 0], [0, 1, 1]])
test("rot: standard grid (expect 4)", r == 4, f"got {r}")
r = rotting_oranges([[2, 1, 1], [0, 1, 1], [1, 0, 1]])
test("rot: isolated fresh orange (expect -1)", r == -1, f"got {r}")
r = rotting_oranges([[0, 2]])
test("rot: no fresh oranges (expect 0)", r == 0, f"got {r}")
r = rotting_oranges([[1, 2]])
test("rot: one step (expect 1)", r == 1, f"got {r}")
r = rotting_oranges([[2, 2], [2, 2]])
test("rot: all already rotten (expect 0)", r == 0, f"got {r}")
r = rotting_oranges([[0]])
test("rot: single empty cell (expect 0)", r == 0, f"got {r}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
